### PR TITLE
introduce show create secondary indexes from table_name

### LINKED ISSUE #74748: sql: add SHOW CREATE INDEXES statement
The output of `SHOW INDEXES` is verbose and not well suited for getting a quick understanding of what indexes exist on a table. The output includes a row for each column (key or stored) in each index. For example:

```
CREATE TABLE t (
    k INT PRIMARY KEY,
    a INT,
    b STRING,
    UNIQUE INDEX (a),
    INDEX (a, b)
);

SHOW INDEXES FROM t;
  table_name | index_name | non_unique | seq_in_index | column_name | direction | storing | implicit
-------------+------------+------------+--------------+-------------+-----------+---------+-----------
  t          | primary    |   false    |            1 | k           | ASC       |  false  |  false
  t          | primary    |   false    |            2 | a           | N/A       |  true   |  false
  t          | primary    |   false    |            3 | b           | N/A       |  true   |  false
  t          | t_a_b_idx  |    true    |            1 | a           | ASC       |  false  |  false
  t          | t_a_b_idx  |    true    |            2 | b           | ASC       |  false  |  false
  t          | t_a_b_idx  |    true    |            3 | k           | ASC       |  false  |   true
  t          | t_a_key    |   false    |            1 | a           | ASC       |  false  |  false
  t          | t_a_key    |   false    |            2 | k           | ASC       |  false  |   true
```

The output of `SHOW CREATE TABLE` presents the indexes in a format that is much easier to understand:

```
SHOW CREATE TABLE t;
  table_name |               create_statement
-------------+------------------------------------------------
  t          | CREATE TABLE public.t (
             |     k INT8 NOT NULL,
             |     a INT8 NULL,
             |     b STRING NULL,
             |     CONSTRAINT "primary" PRIMARY KEY (k ASC),
             |     UNIQUE INDEX t_a_key (a ASC),
             |     INDEX t_a_b_idx (a ASC, b ASC),
             |     FAMILY "primary" (k, a, b)
             | )
```

But `SHOW CREATE TABLE` includes columns, constraints, and families, making it difficult to extract just the index information if that is all that is desired.

A `SHOW CREATE INDEXES FROM <table>` statement could output `CREATE INDEX` statements that represent all the indexes of a table.

For example:

```
SHOW CREATE INDEXES FROM t;
               create_statement
--------------------------------------------
 CREATE UNIQUE INDEX t_a_key ON t (a ASC)
 CREATE INDEX t_a_b_idx ON t (a ASC, b ASC)
```

An open question is whether or not the primary index should be included, and if so, how to represent it.

Jira issue: CRDB-12244

Epic CRDB-17785

### PR DESCRIPTION
**Pull request by arenatlx**
Signed-off-by: arenatlx <tlxarena@gmail.com>

closes https://github.com/cockroachdb/cockroach/issues/74748

Release note (sql change): Introduced a `SHOW SECONDARY INDEXES FROM table` command, which shows all secondary indexes from a table.

### CODE REVIEW

**Review by Xiang-Gu — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79201)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arenatlx and @mgartner)

___
*[pkg/sql/logictest/testdata/logic_test/show_create, line 124 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79201#-N0C3orF5Nx3cQ7Ilv4j:-N0C3orF5Nx3cQ7Ilv4k:b-tppq24) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bedac91a793b85aee584560038dec9b0e2cec81/pkg/sql/logictest/testdata/logic_test/show_create#L124)):*
> ```
> CREATE INDEX c_a_b_idx ON c (a ASC, b ASC)
> CREATE UNIQUE INDEX unique_a ON c (a ASC)
> CREATE INDEX test_idx ON c (b DESC, a ASC)
> ```

Hi, I'm new to the team and now learning to get myself into code reviews. I'm just browsing PRs to learn what other people are doing and ask some questions around. Please don't count solely on me to be the one to give final approval even if I might be the first one to leave a comment.


1. why don't we show the full resolved path for `c` in the response? e.g. CREATE INDEX test_idx ON public.c (b DESC, a ASC) because this is exactly the same command that created this index.

2. What will happen if the index is created implicitly? E.g. if I create a table with a unique column, a unique secondary index will be automatically created. Will we be able to also show this implicitly created secondary index?

3. What will happen for stored columns in the index? What should the response look like?

4. Did we make sure the response is consistent with "SHOW INDEX FROM table" in that they're identical except for the primary index?

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79201)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arenatlx, @mgartner, and @Xiang-Gu)

___
*[pkg/sql/delegate/show_table.go, line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79201#-N0CNrfY6B22e0OBiU0L:-N0CNrfY6B22e0OBiU0M:b-hqjkk5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bedac91a793b85aee584560038dec9b0e2cec81/pkg/sql/delegate/show_table.go#L94)):*
> ```Go
> 	index_basic`
> 
> 	return d.showTableDetails(n.Name, showCreateSecondaryIndexFromTable)
> ```

I think that my preference for the implementation of this would be to add a `create_statement` field to `crdb_internal.table_indexes` and then leverage https://github.com/cockroachdb/cockroach/blob/36d1b608bbabd7cf40bf18db513012653dabbfe9/pkg/sql/catalog/catformat/index.go#L53 internally. If there's concern that that'll make the virtual table more expensive, then I'd like us to refactor the logic of the virtual table out and add a new virtual table which has the added field, maybe `crdb_internal.table_indexes_with_create`. Then the delegate can access that directly. I'd also like us to leverage name resolution via `regclass` if we can so that we can get `O(1)` lookup for a single table.

Concretely, this means:

1) Add a `create_statement` column to `crdb_internal.table_indexes`
2) Change this delegate to:
```
SELECT create_statement FROM crdb_internal.table_indexes WHERE descriptor_id = $1::regclass::int AND index_type != 'primary'
```
___
*[pkg/sql/logictest/testdata/logic_test/show_create, line 124 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79201#-N0C3orF5Nx3cQ7Ilv4j:-N0CMVeu6k9zoZpRvnKV:b-6u0sg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bedac91a793b85aee584560038dec9b0e2cec81/pkg/sql/logictest/testdata/logic_test/show_create#L124)):*
These are all great questions! 

> Did we make sure the response is consistent with "SHOW INDEX FROM table" in that they're identical except for the primary index?

`SHOW INDEX` has a tabular format that doesn't print any syntax like this.

**Comment by arenatlx** [comment]
impressive instructions, thx a lot @ajwerner @Xiang-Gu

**Comment by arenatlx** [comment]
PTAL again thx @ajwerner @Xiang-Gu

**Review by Xiang-Gu — COMMENTED** [review]
Hi @arenatlx Can you take some time to address our comments?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79201#-:-N2XuYa9348rjs5T2Ak7:b3fxt2s)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arenatlx, @mgartner, and @Xiang-Gu)

**Comment by arenatlx** [comment]
> Hi @arenatlx Can you take some time to address our comments?
> 
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79201#-:-N2XuYa9348rjs5T2Ak7:b3fxt2s)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arenatlx, @mgartner, and @Xiang-Gu)

I think i did.
I followed what [ajwerner](https://github.com/ajwerner) said, and for your questions.
1: yes we did, `IndexForDisplay` can output like that.
2: yes we did, unique keyword is also a kind of unique constraint implemented by a unique index. 
3: for a stored column, since `IndexForDisplay` hasn't unfolded them to expression before, we output the delegate name.
4: `IndexForDisplay` has its tabular format quite different from what the index in show-create-table does.

**Review by otan — COMMENTED** [review]
can you fold all these commits into one? `git rebase -i HEAD~9` and doing a fixup on the following 8 commits and then `git push --force` should do the trick.

Please also make sure that one commit has the release note. Something like:

```
Release note (sql change): Introduced a `SHOW SECONDARY INDEXES FROM table`
command, which shows all secondary indexes from a table.
```

File: pkg/sql/delegate/show_table.go [hunk 1]
```diff
@@ -52,6 +54,14 @@ WHERE name = %s
 	return parse(fmt.Sprintf(showCreateQuery, lexbase.EscapeSQLString(n.Name.Object())))
 }
 
+func (d *delegator) delegateShowCreateSecondaryIndexes(n *tree.ShowCreate) (tree.Statement, error) {
+	sqltelemetry.IncrementShowCounter(sqltelemetry.Indexes)
+
+	const showCreateSecondaryIndexFromTable = `SELECT create_statement FROM crdb_internal.table_indexes WHERE descriptor_id = %[2]s::regclass::int AND index_type != 'primary'`
```
Comment (otan) [inline pkg/sql/delegate/show_table.go:60]: i think this doesn't work when you use `SHOW CREATE SECONDARY INDEXES FROM <other_db>.table`. can you try it on a table on another database in the lgoic tests?

to fix this, you'll need to follow how `SHOW CREATE TABLE` references `<database_name>.crdb_internal.table_indexes` which will fetch the correct result

File: pkg/sql/logictest/testdata/logic_test/show_create [hunk 2]
```diff
@@ -140,3 +140,19 @@ CREATE TABLE public.t (
...
+SHOW CREATE SECONDARY INDEXES FROM public.c
+----
+CREATE INDEX c_a_b_idx ON test.public.c (a ASC, b ASC)
+CREATE UNIQUE INDEX unique_a ON test.public.c (a ASC)
+
+statement ok
+CREATE INDEX test_idx ON public.c (b DESC, a ASC);
+
+query T rowsort
+SHOW CREATE SECONDARY INDEXES FROM public.c
```
Comment (otan) [inline pkg/sql/logictest/testdata/logic_test/show_create:154]: can you try this on a table which doesn't exist as well?

**Review by Xiang-Gu — COMMENTED** [review]
I'm sorry that I was not aware of the request and I let it sit for so long. Your changes look good to me! Thank you for contributing to CRDB!

One last thing, I saw you added logic to build the PARTITION BY clause.  Can you add a test case to show a secondary index with PARTITION BY? 

You will need to fix the CI as well!

**Review by Xiang-Gu — APPROVED** [review]

**Comment by arenatlx** [comment]
from the bazel ci log, it's hard to tell what's going on, I have no idea.  =.=  Somebody can help take a look at this?
(even GitHub ci has already been successful before)

**Pull request closed without merging**